text/0000-toolchains.md

```diff
@@ -0,0 +1,155 @@
+- Feature Name: toolchains
+- Start Date: 2018-12-07
+- RFC PR: 
+- Notion Issue: 
+
+# Summary
+[summary]: #summary
+
+This RFC describes a design for **toolchains**, the central unifying concept of Notion.
+
+# Motivation
+[motivation]: #motivation
+
+Notion toolchains attempt to achieve several goals:
+
+- **Reproducible project tooling.** With this design, Notion should make it convenient and reliable for all contributors on a project to get the same exact version of the Node runtime, the package manager, and any package binaries configured for use by `package.json`.
+- **Reconciling the use of npm as a software distribution platform with best practices.** It's considered a best practice for project development to avoid ever installing global packages. But npm is a popular and convenient platform for distributing command-line tools. This design reconciles the tension by isolating _user tools_, which are installed for personal use, making them invisible to JS project scripts, and distinguishing them from project tools.
+- **Install and forget.** This design solves the problem of _tool bitrot_, where a tool stops working because of upgrades to the system Node version. It also avoids the need to reinstall global tools every time a new Node version is provisioned. With Notion, once you install a user tool and get it working, it keeps working unless you deliberately decide to change or uninstall it.
+- **Lightweight cognitive model via statelessness.** This design allows Notion users to generally avoid thinking at all about the version of Node that's currently installed, instead relying on saved configuration to ensure that tools and projects have already declaratively specified their Node platform version.
+
+## User stories
+
+To demonstrate this, here are three representative user stories.
+
+### Project tools: `tsc`
+
+A project maintainer selects a version of TypeScript in `package.json`:
+
+```js
+  ...
+  "dependencies": {
+      "typescript": "^3.2"
+  },
+  "platform": {
+      "node": "10.10.0"
+  }
+  ...
+```
+
+and the lockfile pins TypeScript to version 3.2.2.
+
+Any Notion user who checks out the project gets an identical development experience when working within the project directory tree. Namely, typing
+```
+tsc
+```
+runs version 3.2.2 of the TypeScript compiler, using Node 10.10.0 as the runtime.
+
+### User tools: `surge`
+
+An end user of the [surge.sh](https://surge.sh) service installs their CLI tool that is deployed as an npm package:
+```
+notion install surge
+```
+Assuming the `surge` tool selects Node 11.4.0 in its `"platform"` manifest (it's OK if not; more details below), the CLI tool is installed in the user toolchain and pinned to Node 11.4.0.
```

Comment:
I think you forgot to elaborate what you meant to “below” here. I’m assuming it picks up whatever version the user has already specified for their user toolchain.

Reply:
Ah yes, thanks! Will add.

Reply:
I just added a section on installation. Thanks for catching this oversight!